src/python/pants/backend/terraform/tffmt.py

```diff
@@ -0,0 +1,198 @@
+from pants.core.goals.fmt import FmtResult
+from pants.core.goals.lint import LintResult, LintResults
+from pants.core.util_rules import external_tool
+from pants.core.util_rules.external_tool import DownloadedExternalTool, ExternalToolRequest
+from pants.core.util_rules.source_files import SourceFiles, SourceFilesRequest
+from pants.engine.fs import Digest, MergeDigests
+from pants.engine.internals.selectors import Get, MultiGet
+from pants.engine.platform import Platform
+from pants.engine.process import FallibleProcessResult, Process, ProcessResult
+from pants.engine.rules import collect_rules, rule
+from pants.engine.target import FieldSet
+from pants.engine.unions import UnionRule
+from pants.option.subsystem import Subsystem
+from pants.util.logging import LogLevel
+from pants.util.strutil import pluralize
+
+logger = logging.getLogger(__name__)
+
+
+class TfFmtSubsystem(Subsystem):
+    options_scope = "tffmt"
+    help = """Terraform fmt options."""
+
+    @classmethod
+    def register_options(cls, register):
+        super().register_options(register)
+        register(
+            "--skip",
+            type=bool,
+            default=False,
+            help=(
+                f"Don't use `terraform fmt` when running `{register.bootstrap.pants_bin_name} fmt` and "
+                f"`{register.bootstrap.pants_bin_name} lint`."
+            ),
+        )
+
+
+@dataclass(frozen=True)
+class TffmtFieldSet(FieldSet):
+    required_fields = (TerraformSources,)
+
+    sources: TerraformSources
+
+
+class TffmtRequest(TerraformFmtRequest):
+    field_set_type = TffmtFieldSet
+
+
+@dataclass(frozen=True)
+class SetupRequest:
+    request: TffmtRequest
+    check_only: bool
+
+
+@dataclass(frozen=True)
+class Setup:
+    directory_to_process: Dict[str, Process]
+    original_digest: Digest
+
+
+@rule(level=LogLevel.DEBUG)
+async def setup_terraform_fmt(setup_request: SetupRequest, terraform: TerraformTool) -> Setup:
+    download_terraform_request = Get(
+        DownloadedExternalTool,
+        ExternalToolRequest,
+        terraform.get_request(Platform.current),
+    )
+
+    source_files_request = Get(
+        SourceFiles,
+        SourceFilesRequest(field_set.sources for field_set in setup_request.request.field_sets),
+    )
+
+    downloaded_terraform, source_files = await MultiGet(
+        download_terraform_request, source_files_request
+    )
+
+    source_files_snapshot = (
+        source_files.snapshot
+        if setup_request.request.prior_formatter_result is None
+        else setup_request.request.prior_formatter_result
+    )
+
+    input_digest = await Get(
+        Digest,
+        MergeDigests((source_files_snapshot.digest, downloaded_terraform.digest)),
+    )
+
+    # `terraform fmt` operates on a directory-by-directory basis. First determine the directories in
+    # the snapshot. This does not use `source_files_snapshot.dirs` because that will be empty if the files
+    # are in a single directory.
```

Comment:
Is that expected..? We definitely do depend on `snapshot.dirs` in other places.

Reply:
On Terraform's part or Pants? It is behavior that I had observed so I worked around it.